File: packages/js/test/signing-client.test.ts
import { SigningCosmWasmClient } from "@cosmjs/cosmwasm-stargate";
import type { OfflineSigner } from "@cosmjs/proto-signing";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

import { SafHandleSigningClient } from "../dist/index.js";

// Exercises the write path end-to-end WITHOUT broadcasting: we stub CosmJS so we
// can assert exactly which ExecuteMsg and funds the SDK would submit.

const SENDER = "addr_safro1sender";
const CONTRACT = "addr_safro1contract";

const CONFIG = {
  native_denom: "usaf",
  name_registration_fee_usaf: "50000000",
  dev_module_wallet: "addr_safro1dev",
  governance_admin: "addr_safro1gov",
  reserved_names: ["safro"],
};

let execute: ReturnType<typeof vi.fn>;
let queryContractSmart: ReturnType<typeof vi.fn>;

async function makeClient(): Promise<SafHandleSigningClient> {
  execute = vi.fn().mockResolvedValue({ transactionHash: "DEADBEEF", height: 1 });
  queryContractSmart = vi.fn().mockResolvedValue(CONFIG);

  const fakeCosmwasm = { execute, queryContractSmart };
  vi.spyOn(SigningCosmWasmClient, "connectWithSigner").mockResolvedValue(
    fakeCosmwasm as unknown as SigningCosmWasmClient,
  );

  const signer = {
    getAccounts: async () => [{ address: SENDER, algo: "secp256k1", pubkey: new Uint8Array() }],
  } as unknown as OfflineSigner;

  return SafHandleSigningClient.connectWithSigner("rpc://x", signer, CONTRACT);
}

beforeEach(async () => {
  // fresh mocks per test via makeClient
});

afterEach(() => {
  vi.restoreAllMocks();
});

describe("registerName", () => {
  it("normalizes the name and attaches the exact fee from config", async () => {
    const client = await makeClient();
    await client.registerName("John");

    expect(queryContractSmart).toHaveBeenCalledWith(CONTRACT, { config: {} });
    expect(execute).toHaveBeenCalledWith(
      SENDER,
      CONTRACT,
      { register_name: { name: "john.saf" } },
      "auto",
      undefined,
      [{ denom: "usaf", amount: "50000000" }],
    );
  });

  it("uses an explicit fee override and skips the config query", async () => {
    const client = await makeClient();
    await client.registerName("alice", { fee: { denom: "usaf", amount: "1" } });

    expect(queryContractSmart).not.toHaveBeenCalled();
    expect(execute).toHaveBeenCalledWith(
      SENDER,
      CONTRACT,
      { register_name: { name: "alice.saf" } },
      "auto",
      undefined,
      [{ denom: "usaf", amount: "1" }],
    );
  });

  it("rejects invalid input before any network call", async () => {
    const client = await makeClient();
    await expect(client.registerName("a")).rejects.toMatchObject({ code: "InvalidName" });
    expect(queryContractSmart).not.toHaveBeenCalled();
    expect(execute).not.toHaveBeenCalled();
  });
});

describe("transfer / release (no funds)", () => {
  it("transferName sends normalized name and new owner with no funds", async () => {
    const client = await makeClient();
    await client.transferName("John", "addr_safro1new");

    expect(execute).toHaveBeenCalledWith(
      SENDER,
      CONTRACT,
      { transfer_name: { name: "john.saf", new_owner: "addr_safro1new" } },
      "auto",
      undefined,
      undefined,
    );
  });

  it("releaseName normalizes and sends no funds", async () => {
    const client = await makeClient();
    await client.releaseName("John");

    expect(execute).toHaveBeenCalledWith(
      SENDER,
      CONTRACT,
      { release_name: { name: "john.saf" } },
      "auto",
      undefined,
      undefined,
    );
  });
});

describe("sender", () => {
  it("exposes the signer's first account address", async () => {
    const client = await makeClient();
    expect(client.sender).toBe(SENDER);
  });
});

File: packages/dart/test/signing_client_test.dart
import 'dart:convert';

import 'package:http/testing.dart';
import 'package:http/http.dart' as http;
import 'package:safhandle/safhandle.dart';
import 'package:test/test.dart';

// Exercises the write path end-to-end WITHOUT broadcasting: a fake signer
// captures exactly which ExecuteMsg and funds the SDK would submit, and a fake
// HTTP client serves the contract config. Mirrors
// packages/js/test/signing-client.test.ts.

const sender = 'addr_safro1sender';
const contract = 'addr_safro1contract';

const config = <String, dynamic>{
  'native_denom': 'usaf',
  'name_registration_fee_usaf': '50000000',
  'dev_module_wallet': 'addr_safro1dev',
  'governance_admin': 'addr_safro1gov',
  'reserved_names': <String>['safro'],
};

/// Records the last execute call for assertions.
class RecordingSigner implements SafHandleSigner {
  Map<String, dynamic>? lastMsg;
  List<Coin>? lastFunds;
  String? lastContract;
  String? lastMemo;
  int calls = 0;

  @override
  String get sender => 'addr_safro1sender';

  @override
  Future<ExecuteResult> execute({
    required String contract,
    required Map<String, dynamic> msg,
    List<Coin> funds = const <Coin>[],
    String? memo,
  }) async {
    calls++;
    lastContract = contract;
    lastMsg = msg;
    lastFunds = funds;
    lastMemo = memo;
    return const ExecuteResult(transactionHash: 'DEADBEEF', height: 1);
  }
}

/// A fake REST gateway. Counts config queries so tests can assert whether the
/// fee lookup happened.
({http.Client client, int Function() configQueries}) fakeRest() {
  var configQueries = 0;
  final client = MockClient((http.Request request) async {
    // Path: /cosmwasm/wasm/v1/contract/{addr}/smart/{b64}
    final segments = request.url.pathSegments;
    final encoded = segments.last;
    final decoded =
        jsonDecode(utf8.decode(base64.decode(Uri.decodeComponent(encoded))))
            as Map<String, dynamic>;
    if (decoded.containsKey('config')) {
      configQueries++;
      return http.Response(jsonEncode(<String, dynamic>{'data': config}), 200,
          headers: const <String, String>{'content-type': 'application/json'});
    }
    return http.Response('unexpected query', 400);
  });
  return (client: client, configQueries: () => configQueries);
}

SafHandleSigningClient makeClient(
    RecordingSigner signer, http.Client httpClient) {
  return SafHandleSigningClient(
    signer: signer,
    restEndpoint: 'https://rest.example',
    contractAddress: contract,
    httpClient: httpClient,
  );
}

void main() {
  group('registerName', () {
    test('normalizes the name and attaches the exact fee from config',
        () async {
      final signer = RecordingSigner();
      final rest = fakeRest();
      final client = makeClient(signer, rest.client);

      await client.registerName('John');

      expect(rest.configQueries(), 1);
      expect(signer.lastContract, contract);
      expect(signer.lastMsg, <String, dynamic>{
        'register_name': <String, dynamic>{'name': 'john.saf'},
      });
      expect(signer.lastFunds, <Coin>[const Coin('usaf', '50000000')]);
      expect(signer.lastMemo, isNull);
    });

    test('uses an explicit fee override and skips the config query', () async {
      final signer = RecordingSigner();
      final rest = fakeRest();
      final client = makeClient(signer, rest.client);

      await client.registerName(
        'alice',
        options: const WriteOptions(fee: Coin('usaf', '1')),
      );

      expect(rest.configQueries(), 0);
      expect(signer.lastMsg, <String, dynamic>{
        'register_name': <String, dynamic>{'name': 'alice.saf'},
      });
      expect(signer.lastFunds, <Coin>[const Coin('usaf', '1')]);
    });

    test('rejects invalid input before any network or signer call', () async {
      final signer = RecordingSigner();
      final rest = fakeRest();
      final client = makeClient(signer, rest.client);

      await expectLater(
        client.registerName('a'),
        throwsA(isA<SafHandleError>()
            .having((e) => e.code, 'code', SafHandleErrorCode.invalidName)),
      );
      expect(rest.configQueries(), 0);
      expect(signer.calls, 0);
    });
  });

  group('transfer / release (no funds)', () {
    test('transferName sends normalized name and new owner with no funds',
        () async {
      final signer = RecordingSigner();
      final client = makeClient(signer, fakeRest().client);

      await client.transferName('John', 'addr_safro1new');

      expect(signer.lastMsg, <String, dynamic>{
        'transfer_name': <String, dynamic>{
          'name': 'john.saf',
          'new_owner': 'addr_safro1new',
        },
      });
      expect(signer.lastFunds, isEmpty);
    });

    test('releaseName normalizes and sends no funds', () async {
      final signer = RecordingSigner();
      final client = makeClient(signer, fakeRest().client);

      await client.releaseName('John');

      expect(signer.lastMsg, <String, dynamic>{
        'release_name': <String, dynamic>{'name': 'john.saf'},
      });
      expect(signer.lastFunds, isEmpty);
    });
  });

  group('sender', () {
    test("exposes the signer's address", () {
      final client = makeClient(RecordingSigner(), fakeRest().client);
      expect(client.sender, sender);
    });
  });
}
